### PR TITLE
Partially implemented type propagation

### PR DESCRIPTION
**Pull request by JuliaPoo**
This branch will be continually updated until it is convenient to merge with `ssa_graphs`

1. implements the types for the symbolics relevant to the current guards
2. modifies the dsl and bytecodes.c to encode the type prop rules
3. changes abstract interpreter for impure ops to only track stack pointer changes
4. added type initialisation of symbolics from constants (sym_const_init)
5. added type propagation for pure operations

### CODE REVIEW

File: Python/optimizer_analysis.c [hunk 1]
```diff
@@ -43,6 +44,119 @@ _PyOpcode_isterminal(int opcode)
...
+    // CHECK_STACK_SPACE_TYPE // Environment check
+    INVALID_TYPE = -1
+} _Py_UOpsSymExprTypeEnum;
+
+#define MAX_TYPE 7
+typedef struct {
+    // bitmask of types
+    uint32_t types;
+    // auxillary data for the types
+    uint32_t aux[MAX_TYPE + 1];
```
Comment (Fidget-Spinner) [inline Python/optimizer_analysis.c]: Is this due to the fact that a single thing could be of multiple positive types?

Would it be possible to not have this, and only have something identify as a single positive type?

Comment (JuliaPoo) [reply Python/optimizer_analysis.c]: Nope, its bcuz multiple guards has to pass for some cases inorder to fully eliminate the entire guard region. I dont see whats the disadvantage of doing this either

Comment (Fidget-Spinner) [reply Python/optimizer_analysis.c]: Every time we add a new type, all sym expressions become 32 bits bigger. I would like to save memory somehow. But ok lets go with this for now.

File: Python/optimizer_analysis.c [hunk 2]
```diff
@@ -43,6 +44,119 @@ _PyOpcode_isterminal(int opcode)
...
+symtype_set_type(_Py_UOpsSymType* sym_type, _Py_UOpsSymExprTypeEnum typ, uint32_t aux) 
+{
+    sym_type->types |= 1 << typ;
+    sym_type->aux[typ] = aux;
+}
+
+static void
+symtype_set_from_const(_Py_UOpsSymType* sym_type, PyObject* obj)
+{
+    PyTypeObject *tp = obj->ob_type;
```
Comment (Fidget-Spinner) [inline Python/optimizer_analysis.c]: Usually these are accessed through macros
```suggestion
    PyTypeObject *tp = Py_TYPE(obj);
```

**Review by Fidget-Spinner — COMMENTED** [review]
Just one code nit from me, thank you.

File: Python/optimizer_analysis.c [hunk 3]
```diff
@@ -43,6 +44,119 @@ _PyOpcode_isterminal(int opcode)
...
+typedef struct {
+    // bitmask of types
+    uint32_t types;
+    // auxillary data for the types
+    uint32_t aux[MAX_TYPE + 1];
+} _Py_UOpsSymType;
+
+/*
+static _Py_UOpsSymExprTypeEnum 
+symtype_guard_to_type_enum(int guard_opcde)
```
Comment (Fidget-Spinner) [inline Python/optimizer_analysis.c]:
```suggestion
symtype_guard_to_type_enum(int guard_opcde)
```

File: Python/optimizer_analysis.c [hunk 4]
```diff
@@ -43,6 +44,119 @@ _PyOpcode_isterminal(int opcode)
...
+        case _GUARD_TYPE_VERSION_STORE: return GUARD_TYPE_VERSION_STORE_TYPE;
+        case _GUARD_DORV_VALUES_INST_ATTR_FROM_DICT: return GUARD_DORV_VALUES_INST_ATTR_FROM_DICT_TYPE;
+        case _GUARD_KEYS_VERSION: return GUARD_KEYS_VERSION_TYPE;
+        case _GUARD_TYPE_VERSION: return GUARD_TYPE_VERSION_TYPE;
+        default: Py_UNREACHABLE();
+    }
+}
+
+static bool
+symtype_passes_guard(_Py_UOpsSymType* sym_type, int guard_opcode, uint32_t aux)
```
Comment (Fidget-Spinner) [inline Python/optimizer_analysis.c]:
```suggestion
symtype_passes_guard(_Py_UOpsSymType* sym_type, int guard_opcode, uint32_t aux)
```

File: Python/optimizer_analysis.c [hunk 5]
```diff
@@ -43,6 +44,119 @@ _PyOpcode_isterminal(int opcode)
...
+        case GUARD_TYPE_VERSION_STORE_TYPE:
+        case GUARD_TYPE_VERSION_TYPE:
+        case GUARD_KEYS_VERSION_TYPE: return aux == sym_type->aux[typ];
+        default: Py_UNREACHABLE();
+    }
+}
+*/
+
+static void
+symtype_set_type(_Py_UOpsSymType* sym_type, _Py_UOpsSymExprTypeEnum typ, uint32_t aux)
```
Comment (Fidget-Spinner) [inline Python/optimizer_analysis.c]:
```suggestion
symtype_set_type(_Py_UOpsSymType* sym_type, _Py_UOpsSymExprTypeEnum typ, uint32_t aux)
```

File: Python/optimizer_analysis.c [hunk 6]
```diff
@@ -43,6 +44,119 @@ _PyOpcode_isterminal(int opcode)
...
+}
+
+static void
+symtype_set_from_const(_Py_UOpsSymType* sym_type, PyObject* obj)
+{
+    PyTypeObject *tp = obj->ob_type;
+
+    if (tp == &PyLong_Type) {
+        sym_type->types |= 1 << PYINT_TYPE;
+    }
```
Comment (Fidget-Spinner) [inline Python/optimizer_analysis.c]:
```suggestion
    }
```

File: Python/optimizer_analysis.c [hunk 7]
```diff
@@ -43,6 +44,119 @@ _PyOpcode_isterminal(int opcode)
...
+        sym_type->types |= 1 << PYUNICODE_TYPE;
+    }
+
+    if (tp->tp_flags & Py_TPFLAGS_MANAGED_DICT) {
+        PyDictOrValues *dorv = _PyObject_DictOrValuesPointer(obj);
+
+        if (_PyDictOrValues_IsValues(*dorv) ||
+            _PyObject_MakeInstanceAttributesFromDict(obj, dorv)) {
+            sym_type->types |= 1 << GUARD_DORV_VALUES_INST_ATTR_FROM_DICT_TYPE;
+
```
Comment (Fidget-Spinner) [inline Python/optimizer_analysis.c]:
```suggestion

```

**Review by Fidget-Spinner — COMMENTED** [review]
Also bunch of whitespaces are failing the CI

File: Tools/cases_generator/stacking.py [hunk 8]
```diff
@@ -185,10 +186,12 @@ def __init__(
         self.final_offset = pred.final_offset.clone() if pred else StackOffset()
         for eff in reversed(instr.input_effects):
             self.final_offset.deeper(eff)
-            self.peeks.append(StackItem(offset=self.final_offset.clone(), effect=eff))
+            self.peeks.append(
+                StackItem(offset=self.final_offset.clone(), effect=eff))
```
Comment (Fidget-Spinner) [inline Tools/cases_generator/stacking.py]: Lets reduce the diff
```suggestion
            self.peeks.append(StackItem(offset=self.final_offset.clone(), effect=eff))
```

File: Tools/cases_generator/stacking.py [hunk 9]
```diff
@@ -185,10 +186,12 @@ def __init__(
...
         self.min_offset = self.final_offset.clone()
         for eff in instr.output_effects:
-            self.pokes.append(StackItem(offset=self.final_offset.clone(), effect=eff))
+            self.pokes.append(
+                StackItem(offset=self.final_offset.clone(), effect=eff))
```
Comment (Fidget-Spinner) [inline Tools/cases_generator/stacking.py]:
```suggestion
            self.pokes.append(StackItem(offset=self.final_offset.clone(), effect=eff))
```

File: Tools/cases_generator/parsing.py [hunk 10]
```diff
@@ -262,14 +270,24 @@ def cache_effect(self) -> CacheEffect | None:
 
     @contextual
     def stack_effect(self) -> StackEffect | None:
-        #   IDENTIFIER [':' IDENTIFIER [TIMES]] ['if' '(' expression ')']
+        #   IDENTIFIER [':' [IDENTIFIER [TIMES]] ['~' '(' IDENTIFIER [',' IDENTIFIER] ')']] ['if' '(' expression ')']
```
Comment (Fidget-Spinner) [inline Tools/cases_generator/parsing.py]: The `MINUS` isn't documented.

File: Tools/cases_generator/stacking.py (hunk 8 above)
Comment (JuliaPoo) [reply Tools/cases_generator/stacking.py]: Ah I didn't realise but at some point during the buggy vscode fiasco it ran the autoformatter

File: Tools/cases_generator/stacking.py [hunk 11]
```diff
@@ -509,11 +510,143 @@ def write_single_instr_for_abstract_interp(instr: Instruction, out: Formatter) -
...
+            copy = dataclasses.replace(peek.effect)
+            copy.name = f"__{copy.name}"
+            out.assign(copy, peek.as_stack_effect())
+
+        # Adjust stack
+        if mgr is managers[-1]:
+            out.stack_adjust(mgr.final_offset.deep, mgr.final_offset.high)
+            # Use clone() since adjust_inverse() mutates final_offset
+            mgr.adjust_inverse(mgr.final_offset.clone())
+
```
Comment (Fidget-Spinner) [inline Tools/cases_generator/stacking.py]:
```suggestion
            mgr.adjust_inverse(mgr.final_offset.clone())

```

File: Tools/cases_generator/stacking.py [hunk 12]
```diff
@@ -511,11 +510,143 @@ def write_single_instr_for_abstract_interp(instr: Instruction, out: Formatter) -
...
```
Comment (Fidget-Spinner) [inline Tools/cases_generator/stacking.py]:
```suggestion

```

**Review by Fidget-Spinner — APPROVED** [review]

**Comment by Fidget-Spinner** [comment]
Thanks for all your work on this.

File: Python/optimizer_analysis.c [hunk 13]
```diff
@@ -44,13 +44,93 @@ _PyOpcode_isterminal(int opcode)
...
+#define MAX_TYPE_WITH_AUX 1
+typedef struct {
+    // bitmask of types
+    uint32_t types;
+    // auxillary data for the types
+    uint32_t aux[MAX_TYPE_WITH_AUX + 1];
+} _Py_UOpsSymType;
+
+static void
+symtype_set_type(_Py_UOpsSymType* sym_type, _Py_UOpsSymExprTypeEnum typ, uint32_t aux)
```
Comment (Fidget-Spinner) [inline Python/optimizer_analysis.c]:
```suggestion
symtype_set_type(_Py_UOpsSymType *sym_type, _Py_UOpsSymExprTypeEnum typ, uint32_t aux)
```

File: Python/optimizer_analysis.c [hunk 14]
```diff
@@ -44,13 +44,93 @@ _PyOpcode_isterminal(int opcode)
...
+{
+    sym_type->types |= 1 << typ;
+    if (typ <= MAX_TYPE_WITH_AUX) {
+        sym_type->aux[typ] = aux;
+    }
+}
+
+static void
+symtype_set_from_const(_Py_UOpsSymType* sym_type, PyObject* obj)
```
Comment (Fidget-Spinner) [inline Python/optimizer_analysis.c]:
```suggestion
symtype_set_from_const(_Py_UOpsSymType *sym_type, PyObject *obj)
```

**Review by Fidget-Spinner — COMMENTED** [review]
PEP 7 nits

**Pull request merged**